### ISSUE TITLE
False positive "has a required though undefined property" on "allOf"

### ISSUE THREAD

**Issue by ili101** [42Crunch/vscode-openapi#33]
```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "test",
        "version": "0.0.1"
    },
    "servers": [
        {
            "url": "https://api.server.test/v1"
        }
    ],
    "components": {
        "schemas": {
            "Param1": {
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "p1": {
                        "type": "string"
                    }
                }
            },
            "Param2": {
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "p2": {
                        "type": "integer"
                    }
                }
            }
        }
    },
    "paths": {
        "/test1": {
            "post": {
                "description": "description test 1",
                "requestBody": {
                    "description": "description body test 1",
                    "required": true,
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object",
                                "additionalProperties": false,
                                "description": "description schema test 1",
                                "required": [
                                    "p1"
                                ],
                                "allOf": [
                                    {
                                        "$ref": "#/components/schemas/Param1"
                                    },
                                    {
                                        "$ref": "#/components/schemas/Param2"
                                    }
                                ]
                            }
                        }
                    }
                },
                "responses": {
                    "200": {}
                }
            }
        }
    }
}
```

Error
```json
{
	"resource": "Untitled-1",
	"owner": "openapi-audit",
	"code": "",
	"severity": 8,
	"message": "Schema '/paths/~1test1/post/requestBody/content/application~1json/schema' has a required though undefined property 'p1' ",
	"startLineNumber": 43,
	"startColumn": 29,
	"endLineNumber": 43,
	"endColumn": 40
}
```

**Comment by ak1394** [comment]
Thanks for reporting, will investigate!

**Comment by ak1394** [comment]
So, after a bit of investigation it looks that the reported error is indeed correct, and your use "allOf" does not work as you would expected. In your example you have `"additionalProperties": false`  and because of that "allOf" will never be checked.

See here for details: https://json-schema.org/understanding-json-schema/reference/combining.html in the section where it mentions "additionalProperties".

**Comment by ili101** [comment]
@ak1394 thank you but it looks like this is not the problem, I removed `"additionalProperties": false` everywhere and I get the same error (note that it works in Swagger UI and AutoRest)
```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "test",
        "version": "0.0.1"
    },
    "servers": [
        {
            "url": "https://api.server.test/v1"
        }
    ],
    "components": {
        "schemas": {
            "Param1": {
                "type": "object",
                "properties": {
                    "p1": {
                        "type": "string"
                    }
                }
            },
            "Param2": {
                "type": "object",
                "properties": {
                    "p2": {
                        "type": "integer"
                    }
                }
            }
        },
        "securitySchemes": {
            "BasicAuth": {
                "type": "http",
                "scheme": "basic"
            }
        }
    },
    "paths": {
        "/test1": {
            "post": {
                "description": "description test 1",
                "requestBody": {
                    "description": "description body test 1",
                    "required": true,
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object",
                                "description": "description schema test 1",
                                "required": [
                                    "p1"
                                ],
                                "allOf": [
                                    {
                                        "$ref": "#/components/schemas/Param1"
                                    },
                                    {
                                        "$ref": "#/components/schemas/Param2"
                                    }
                                ]
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "re"
                    }
                }
            }
        }
    },
    "security": [
        {
            "BasicAuth": []
        }
    ]
}
```

**Comment by ili101** [comment]
Hi
I did some searching and found that this is a valid syntax in OpenAPI 3.0
From OpenAPI-Specification GitHub:
https://github.com/OAI/OpenAPI-Specification/issues/1608
https://github.com/OAI/OpenAPI-Specification/issues/1428
Fix made for the same problem in another validator tool:
https://github.com/p1c2u/openapi-spec-validator/pull/28
Please reopen this issue and fix
Thank you

**Comment by ak1394** [comment]
Thanks for investigating the issue further. I've re-opened the issue and we re-check it on our end.

**Comment by unexpectedShr3k** [comment]
@ak1394 so in the link you shared, it is a recognized shortcoming and we should expect the security audit to always fail as per its version?

**Comment by ak1394** [comment]
We have reopened the issue, as it looks there is some merit in re-investigating it, but we haven't done it yet.

**Comment by peteraritchie** [comment]
See also https://github.com/OAI/OpenAPI-Specification/issues/1590#issuecomment-391519018 that confirms the required array and allOf merges with the required within the reference.

**Comment by peteraritchie** [comment]
a _short_ example where this is causing noise for me:
```yaml
    simple: # Schema has a required but undefined property 'TypeIdentifier' 
      required: [TypeIdentifier]
      allOf:
        - $ref: "#/components/schemas/detail"

    complex: # Schema has a required but undefined property 'TypeIdentifier' 
      required:
        - TypeIdentifier
      allOf:
        - properties:
            TypeIdentifier:
              type: string
              enum:
                - Simple
                - Complex
            name:
              type: string
              pattern: '^example-[0-9a-z]+$'
              maxLength: 32
        - $ref: "#/components/schemas/other-detail"

    detail:
      properties:
        TypeIdentifier:
          type: string
          enum:
            - Simple
            - Complex

    other-detail:
      type: string
      pattern: '^example-[0-9a-z]+$'
      maxLength: 32
```

Entire spec here: https://gist.github.com/peteraritchie/2a97de472561f72e1a30b54edce2c9ee

**Comment by ak1394** [comment]
Thanks for feedback and example! We're looking at it, and I'll update the issue once I have more info.

**Comment by ak1394** [comment]
Fixed in version 4.8.0
